src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/PipelineExtensions.cs

```diff
@@ -68,7 +70,9 @@ public static ValueTask<ArraySegment<byte>> PeekAsync(this IPipeReader pipelineR
                 }
                 finally
                 {
-                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.Start);
+                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.IsEmpty
+                        ? result.Buffer.End
+                        : result.Buffer.Start);
                 }
 
                 readingTask = pipelineReader.ReadAsync();
```

Comment:
Passing `End` to both `consumed` and `examined` would break the idea of "peek" though.

`ConsumedBytes` calls `Advance`, but if we consume here then that call would double consume bytes from the input.